### PR TITLE
feat: Browser to Browser

### PR DESCRIPTION
**Pull request by ckousik**
Adds the browser to browser transport. Supersedes #88 

- [X] Add SDP protobufs.
- [X] Connection receiver initiates the RTCPeerConnection (when connecting from A to B, B will create the SDP offer after A establishes the relayed connection).
- [x] Add tests for protocol handler.
- [x] Add examples to CI.

### CODE REVIEW

**Comment by ckousik** [comment]
@ddimaria Still waiting on resolution for protocol name and code https://github.com/libp2p/specs/pull/497#issuecomment-1383635744

**Comment by mxinden** [comment]
> @ddimaria Still waiting on resolution for protocol name and code [libp2p/specs#497 (comment)](https://github.com/libp2p/specs/pull/497#issuecomment-1383635744)

@ckousik do I understand correctly that these discussions are not blocking this pull request? Or is all other work done?

**Comment by ddimaria** [comment]
> > @ddimaria Still waiting on resolution for protocol name and code [libp2p/specs#497 (comment)](https://github.com/libp2p/specs/pull/497#issuecomment-1383635744)
> 
> @ckousik do I understand correctly that these discussions are not blocking this pull request? Or is all other work done?

@mxinden Chinmay is out for the rest of the week.  I can confirm that the spec-related decisions are not blocking this PR.  I plan on adding a browser-to-browser example to the `/examples` directory and adding that to CI, as well as internally reviewing the code today/tomorrow.  The goal is to have it ready for review sometime tomorrow.

File: src/peer_transport/transport.ts
Comment (ckousik) [inline src/peer_transport/transport.ts:122]: Consider a case where `A` is connecting to `B` using `webrtc-direct`. `A` and `B` both complete SDP exchange and their PeerConnections transition to the `connected` state. For the ICE connection, `A` is the controlled peer and `B` is the controlling peer. With reference to the description [here](https://datatracker.ietf.org/doc/html/rfc8445#section-7.2.5.3.4), `A` moves to the connected state before `B`. This allows `A` to create a stream on the new connection before `B` has a chance to set up the datachannel callbacks.

**Comment by MarcoPolo** [comment]
- [ ] todo We need to add CI testing for this example

File: src/peer_transport/listener.ts
Comment (MarcoPolo) [inline src/peer_transport/listener.ts:21]: Do we need this? Aren't we listening already via some relay?

File: src/peer_transport/transport.ts
Comment (MarcoPolo) [inline src/peer_transport/transport.ts:119]: Do we need to close this? If the above errors, who closes this?

File: src/peer_transport/transport.ts
Comment (MarcoPolo) [inline src/peer_transport/transport.ts:149]: Can we move this close to above the `const conn =` line so it's similar to the `dial` method, please?

**Review by MarcoPolo — COMMENTED** [review]
Minor nits, but overall looks good to me. I'll leave the final approval to @achingbrain 

Great work @ckousik :)

File: examples/go-libp2p-server/relay/main.go
Comment (ckousik) [reply examples/go-libp2p-server/relay/main.go:18]: We haven't landed support for circuit v2 in js yet

File: src/muxer.ts
Comment (ckousik) [reply src/muxer.ts:17]: Sorry, this should be removed.

File: src/peer_transport/transport.ts
Comment (ckousik) [reply src/peer_transport/transport.ts:119]: Changed so that either side of the stream will reset if there is an error in processing the SDP. Also, stream will be closed after a connection is successfully upgraded.

File: src/muxer.ts
Comment (ckousik) [reply src/muxer.ts:118]: `s.stat.timeline.close != null` would filter out all streams that are open, so that condition needs to be inverted. Other than that, I'm testing this change.

File: src/muxer.ts
Comment (ckousik) [reply src/muxer.ts:116]: Thanks for pointing this out. Apparently, `onIncomingStream` is also queued async https://github.com/libp2p/js-libp2p/blob/master/src/upgrader.ts#L375

File: examples/browser-to-browser/index.js
Comment (ckousik) [reply examples/browser-to-browser/index.js:88]: 276 is the protoCode for circuit-relay. This is just used to differentiate between dialing a relay peer to start listening vs dialing a peer over relay to create an echo stream.

File: src/peer_transport/listener.ts
Comment (ckousik) [reply src/peer_transport/listener.ts:21]: Probably not, but the rationale to add this is as follows:
1. Since webrtc-direct is a protocol, we can advertise support over any transport.
2. For running on node, we can use `node-webrtc` to replace all instances of `RTCPeerConnection` (webrtc-star and webrtc-direct currently do this).

Although, an argument can be made that any listener can support webrtc-direct since it is only implemented as a protocol.

**Comment by ddimaria** [comment]
@ckousik the browser-to-browser example is now being tested in playwright and has been added to CI (confirmed it's working).  Do we just need to turn off the [Firefox test](https://github.com/libp2p/js-libp2p-webrtc/actions/runs/4247903065/jobs/7386477354)?

File: examples/go-libp2p-server/relay/main.go
Comment (mxinden) [reply examples/go-libp2p-server/relay/main.go:18]: For the record, tracking issue https://github.com/libp2p/js-libp2p/issues/1029 and pull request https://github.com/libp2p/js-libp2p/pull/1533.

**Comment by p-shahi** [comment]
Status update:
Based on 2023-02-28 meeting, we will merge this PR with @MarcoPolo 's approval and have @achingbrain do a thorough review at a later time.
The necessary Circuit Relay v2 pr in js-libp2p was merged yesterday: https://github.com/libp2p/js-libp2p/pull/1533

The changes needed before merging is incorporating js-multiaddr `webrtc-w3c` (instead of `webrtc-direct`): https://github.com/multiformats/js-multiaddr/pull/309

**Comment by achingbrain** [comment]
CI is failing?  Is this expected?

**Comment by ckousik** [comment]
> CI is failing? Is this expected?

@achingbrain Looks like a Go version update causing the build to fail. The failure on Firefox is expected because it doesn't support local candidates for ICE negotiation.

**Comment by MarcoPolo** [comment]
> The failure on Firefox is expected because it doesn't support local candidates for ICE negotiation.

Is there a Firefox issue somewhere? Where can I learn more? Does it work on Safari?

**Comment by ckousik** [comment]
@MarcoPolo https://mediasoup.discourse.group/t/firefox-ice-failed-add-a-stun-server-and-see-about-webrtc-for-more-details/805/3 and https://bugzilla.mozilla.org/show_bug.cgi?id=1672145

**Comment by MarcoPolo** [comment]
From the linked bugzilla page:
> Not a super high priority because this is localhost only.

Ah I missed that this is only happens on localhost addresses

**Comment by ddimaria** [comment]
@MarcoPolo @achingbrain @p-shahi CI is now passing (thanks @ckousik 🙏).  We had to skip a single test for a single browser (Firefox) due to the browser's limitations.  Is there anything else outstanding?

**Comment by p-shahi** [comment]
Nothing from this PR I believe, just waiting for upstream discussion to resolve before merging: https://github.com/libp2p/js-libp2p-webrtc/pull/90#issuecomment-1458369439

File: src/peer_transport/listener.ts
Comment (MarcoPolo) [inline src/peer_transport/listener.ts:22]: THis is a bit confusing. Is there a better way to write this?

File: examples/go-libp2p-server/relay/main.go
Comment (MarcoPolo) [reply examples/go-libp2p-server/relay/main.go:18]: These are in now, can we update this?

File: src/index.ts
Comment (MarcoPolo) [inline src/index.ts:10]: This param should be optional if an empty obj is valid.

File: src/peer_transport/handler.ts
Comment (MarcoPolo) [inline src/peer_transport/handler.ts:5]: Shouldn't this end in `.js`? Otherwise this will fail to import in Node. Similar
comments for other imports.

**Comment by MarcoPolo** [comment]
We probably want to let users pass in their own stun server to the RTCPeerConnection constructor, right?

**Comment by ckousik** [comment]
> We probably want to let users pass in their own stun server to the RTCPeerConnection constructor, right?

The init object for the transport can have an optional RTCConfiguration object. The STUN server can be passed into this object. https://github.com/libp2p/js-libp2p-webrtc/pull/90/files#diff-ae86b5a052f8a01ec57238e9da2eca51b26220cf10e893315132f38887c13106R25

**Comment by ddimaria** [comment]
@p-shahi @MarcoPolo @achingbrain @ckousik I don't see any outstanding/unresolved comments.  Is there anything blocking this PR from being merged?

File: examples/browser-to-browser/README.md
Comment (p-shahi) [inline examples/browser-to-browser/README.md:1]:
```suggestion
# js-libp2p-webrtc Browser to Browser
```
I wonder if we should rename this to private-to-private or private browser-to-private browser to be better in line with current rename. Let's discuss in standup tomorrow

Comment (ddimaria) [reply examples/browser-to-browser/README.md:1]: I could see both sides of this. If we're looking for this example's name to have parity with the spec, then renaming makes sense.  On the other hand, the example is a browser connecting to a browser, so the naming is aligned with the example.  I don't have a strong opinion either way.

File: examples/browser-to-browser/index.js
Comment (SgtPooki) [reply examples/browser-to-browser/index.js]: is there a library where we can ensure the protocode is more self documenting? (similar to codes in dag-json, dag-cbor, etc.. from ipld) we shouldn't be throwing magic numbers around

File: src/peer_transport/transport.ts
Comment (p-shahi) [inline src/peer_transport/transport.ts:20]: Needs to be addressed

**Review by p-shahi — CHANGES_REQUESTED** [review]
> > This is currently blocked by [multiformats/multicodec#318](https://github.com/multiformats/multicodec/pull/318), [multiformats/multiaddr#151](https://github.com/multiformats/multiaddr/pull/151), and [multiformats/js-multiaddr#314](https://github.com/multiformats/js-multiaddr/pull/314)
> > ETA for those three PRs is tomorrow before the LBL meeting
> > I assume @MarcoPolo will do a pass and review after which (review addressed, 👍 given) it may be good to merge.
> 
> @p-shahi those PRs are now either closed or merged. Is there anything outstanding in this PR that you're aware of?

@ddimaria I think we should address these comments before merging. @MarcoPolo 's going to do another pass as well.

File: src/peer_transport/handler.ts
Comment (MarcoPolo) [inline src/peer_transport/handler.ts:80]: Where do we close the channel?

File: src/peer_transport/handler.ts
Comment (MarcoPolo) [inline src/peer_transport/handler.ts:90]: Can we rename this to "initiateConnection" instead? It makes it clear that the caller is the initiator

File: src/peer_transport/handler.ts
Comment (MarcoPolo) [inline src/peer_transport/handler.ts:29]: I thought it was a bug to send an ICE candidate before an answer (there's a test case for that), why are we setting the callback here rather than after we send the answer?

File: src/peer_transport/listener.ts
Comment (MarcoPolo) [inline src/peer_transport/listener.ts:35]: shouldn't we be encapsulating `/webrtc` here?

File: src/peer_transport/transport.ts
Comment (MarcoPolo) [inline src/peer_transport/transport.ts:24]: We should rename these things to just WebRTC so they match the multiaddr code.

File: src/peer_transport/transport.ts
Comment (MarcoPolo) [inline src/peer_transport/transport.ts:44]: when is this set? I think it will always return false

File: src/peer_transport/transport.ts
Comment (MarcoPolo) [inline src/peer_transport/transport.ts:48]: should we set this._started after this?

File: examples/browser-to-browser/README.md
Comment (MarcoPolo) [inline examples/browser-to-browser/README.md:16]: This doesn't work?

```
❯ npm run node-relay

> js-libp2p-webrtc-browser-to-server@1.0.0 node-relay
> node relay.js

node:internal/modules/cjs/loader:998
  throw err;
  ^

Error: Cannot find module '/Users/marco/pl/libp2p/js-libp2p-webrtc/examples/browser-to-browser/relay.js'
```

File: src/peer_transport/transport.ts
Comment (MarcoPolo) [inline src/peer_transport/transport.ts:34]: who sets this?

File: examples/browser-to-browser/package.json
Comment (MarcoPolo) [inline examples/browser-to-browser/package.json:12]: We should add webkit, but this will fail until
https://github.com/libp2p/js-libp2p/pull/1627 is fixed.

File: examples/browser-to-browser/README.md
Comment (MarcoPolo) [inline examples/browser-to-browser/README.md:32]: I only see "dialing" but I never see a listening on message. Are you able to get
this working? It fails for me on safari, firefox, and chromium (Brave)

**Review by MarcoPolo — CHANGES_REQUESTED** [review]
I wasn't able to get this working locally and CI is failing.

**Comment by ckousik** [comment]
@MarcoPolo The Go relay has been scrapped in favor of a js relay for tests due to https://github.com/libp2p/js-libp2p/issues/1637. This is why you are not seeing the `Listening` message even though the circuit is established. My priority for now is to get the example tests running for browser-to-server.

File: examples/go-libp2p-server/relay/main.go
Comment (ckousik) [reply examples/go-libp2p-server/relay/main.go]: This has been removed

File: examples/go-libp2p-server/main.go
Comment (ckousik) [reply examples/go-libp2p-server/main.go]: Fixed. This was commented out to allow testing the relay, which can depend on a released version of go-libp2p, while the server has to depend on a fork.

File: examples/go-libp2p-server/relay/main.go
Comment (ckousik) [reply examples/go-libp2p-server/relay/main.go]: I think I took this from the example, but no real reason. The file has been removed.

File: src/peer_transport/transport.ts
Comment (ckousik) [reply src/peer_transport/transport.ts:18]: Sure, that had not landed when I wrote this.

File: examples/browser-to-browser/index.js
Comment (ckousik) [reply examples/browser-to-browser/index.js:70]: The webrtc library does not add any new multiaddresses to the peerStore. Given the discussion regarding multiaddresses, I feel it should add a new mulitaddress.

File: test/peer.browser.spec.ts
Comment (ckousik) [reply test/peer.browser.spec.ts]: It does. Firefox is odd since it did not support the `connectionState` instance property. Refer [here](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/connectionState#browser_compatibility). This has been fixed.

File: src/peer_transport/handler.ts
Comment (ckousik) [reply src/peer_transport/handler.ts:29]: Because firefox will start gathering candidates immediately after we create the offer. This is still a valid flow, since the remote will receive an offer followed by candidates.

File: src/peer_transport/listener.ts
Comment (ckousik) [reply src/peer_transport/listener.ts:35]: I'm assuming the multiaddr `ma` contains `/webrtc`, since the transport will filter any multiaddrs that do not  contain it.

**Comment by ddimaria** [comment]
CI is now passing.  Thanks @ckousik!

File: examples/browser-to-browser/package.json
Comment (ddimaria) [reply examples/browser-to-browser/package.json:11]: Is this a blocker @MarcoPolo or a new issue to work once https://github.com/libp2p/js-libp2p/pull/1627 is completed?

File: src/peer_transport/handler.ts
Comment (ckousik) [reply src/peer_transport/handler.ts:66]: Sorry, forgot to delete this comment from a previous implementation. Earlier, the handler was creating a channel, and was therefore responsible for closing it. Now the channel is created by the initiator. The purpose of creating a dummy channel is to add a WebRTC component for which candidates should be gathered.

File: src/peer_transport/transport.ts
Comment (ckousik) [reply src/peer_transport/transport.ts]: Remaining from an older implementation. This field is included in `CreateListenerOptions`. It wouldn't work here since all listening is done by the transport itself over the protocol.

File: src/muxer.ts
Comment (ckousik) [reply src/muxer.ts:29]: There can be a maximum of 2^16 - 1 datachannels. This buffer would only hold references to objects. Please note that a PeerConnection could also initiate 2^16 datachannels after the handshake completes. Since the PeerConnection only notifies the user that a datachannel has been created by the remote using an `ondatachannel` callback.

File: src/peer_transport/util.ts
Comment (ckousik) [reply src/peer_transport/util.ts]: No, I had added it to the empty file so that TS would treat it as a module.

File: src/stream.ts
Comment (ckousik) [reply src/stream.ts:456]: Addressed.

File: examples/browser-to-browser/package.json
Comment (MarcoPolo) [reply examples/browser-to-browser/package.json:11]: Not a blocker

File: src/peer_transport/transport.ts
Comment (MarcoPolo) [reply src/peer_transport/transport.ts]: I pushed a fix.

File: src/peer_transport/handler.ts
Comment (MarcoPolo) [reply src/peer_transport/handler.ts:29]: Isn't it racey then? Couldn't this write to the stream before we write our answer?

Comment (ckousik) [reply src/peer_transport/handler.ts:29]: It's not a  problem in the firefox case,  as it seems the candidates for the peer that generates the answer are gathered after the local description is set, i.e the answer is generated and set. I can add a deferred promise to alleviate the race  condition.

Comment (ckousik) [reply src/peer_transport/handler.ts:31]: Fixed now

File: examples/browser-to-browser/README.md
Comment (maschad) [inline examples/browser-to-browser/README.md:16]: In order to run this example, the `@libp2p/webrtc` package must be built as well, so before running `npm run start` here, one must run `npm run build`  in the root directory then `cd` into `examples/browser-to-browser` and then they can  run 
```shell 
npm i && npm run start
```

File: examples/browser-to-browser/index.js
Comment (MarcoPolo) [reply examples/browser-to-browser/index.js:68]: Why do you need to remove an address?

**Review by MarcoPolo — COMMENTED** [review]
Some small, nits. I'll fix up anything left and merge this soon. 🎉

File: src/peer_transport/transport.ts
Comment (MarcoPolo) [reply src/peer_transport/transport.ts:153]: Asked synchronously. Nope

File: examples/browser-to-server/tests/test.js
Comment (p-shahi) [reply examples/browser-to-server/tests/test.js:94]: > TODO (me or @prithvi) can we make an issue for this to remember to come back to it?

Created this https://github.com/libp2p/js-libp2p-webrtc/issues/112

Comment (MarcoPolo) [reply examples/browser-to-server/tests/test.js:94]: Fixed

**Review by MarcoPolo — APPROVED** [review]

File: examples/browser-to-browser/index.js
Comment (MarcoPolo) [reply examples/browser-to-browser/index.js:68]: https://github.com/libp2p/js-libp2p-webrtc/issues/113

**Comment by MarcoPolo** [comment]
Merging this since I've given it a review. I'm sure there's more things to fix, but I'd rather do those incrementally from now on. @achingbrain feel to review this at a later time

**Pull request merged**
